src/java/org/apache/cassandra/auth/CassandraAuthorizer.java

```diff
@@ -83,18 +86,37 @@ public Set<Permission> authorize(AuthenticatedUser user, IResource resource)
         }
     }
 
-    public void grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
+    public Set<Permission> grant(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource grantee)
     throws RequestValidationException, RequestExecutionException
     {
-        modifyRolePermissions(permissions, resource, grantee, "+");
-        addLookupEntry(resource, grantee);
+        String roleName = escape(grantee.getRoleName());
+        String resourceName = escape(resource.getName());
+        Set<Permission> existingPermissions = getExistingPermissions(roleName, resourceName, permissions);
+        Set<Permission> nonExistingPermissions = Sets.difference(permissions, existingPermissions);
+
+        if (!nonExistingPermissions.isEmpty())
+        {
+            modifyRolePermissions(nonExistingPermissions, resource, grantee, "+");
+            addLookupEntry(resource, grantee);
+        }
+
+        return nonExistingPermissions;
     }
 
-    public void revoke(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource revokee)
+    public Set<Permission> revoke(AuthenticatedUser performer, Set<Permission> permissions, IResource resource, RoleResource revokee)
     throws RequestValidationException, RequestExecutionException
     {
-        modifyRolePermissions(permissions, resource, revokee, "-");
-        removeLookupEntry(resource, revokee);
+        String roleName = escape(revokee.getRoleName());
```

Comment:
why escaping instead of binding?

Reply:
Good question. I've went through the code and the usage of escape is matches what is done in the remaining of the file. 
In `CassandraAuthorizer.grant()` we're escaping `roleName` and `resourceName` and then passing them to: `CassandraAuthorizer.getExistingPermissions()`, `CassandraAuthorizer.modifyRolePermissions()`, and `CassandraAuthorizer.addLookupEntry()`. The two later calls, do `escape` internally only the first one doesn't - I wonder if that's the reason we do it here.

I have found some tests that justify this escaping. Those can be found in: `test/unit/org/apache/cassandra/cql3/validation/miscellaneous/RoleSyntaxTest.java`.